core/services/relay/evm/mercury/v2/data_source.go

```diff
@@ -0,0 +1,199 @@
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+
+	pkgerrors "github.com/pkg/errors"
+
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	relaymercury "github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury"
+	relaymercuryv2 "github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury/v2"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocrcommon"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury/types"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+type Runner interface {
+	ExecuteRun(ctx context.Context, spec pipeline.Spec, vars pipeline.Vars, l logger.Logger) (run pipeline.Run, trrs pipeline.TaskRunResults, err error)
+}
+
+type LatestReportFetcher interface {
+	LatestPrice(ctx context.Context, feedID [32]byte) (*big.Int, error)
+	LatestTimestamp(context.Context) (uint32, error)
+}
+
+type datasource struct {
+	pipelineRunner Runner
+	jb             job.Job
+	spec           pipeline.Spec
+	feedID         types.FeedID
+	lggr           logger.Logger
+	runResults     chan<- pipeline.Run
+
+	fetcher      LatestReportFetcher
+	linkFeedID   types.FeedID
+	nativeFeedID types.FeedID
+
+	mu sync.RWMutex
+
+	chEnhancedTelem chan<- ocrcommon.EnhancedTelemetryMercuryData
+}
+
+var _ relaymercuryv2.DataSource = &datasource{}
+
+var maxInt192 = new(big.Int).Exp(big.NewInt(2), big.NewInt(191), nil)
+
+func NewDataSource(pr pipeline.Runner, jb job.Job, spec pipeline.Spec, feedID types.FeedID, lggr logger.Logger, rr chan pipeline.Run, enhancedTelemChan chan ocrcommon.EnhancedTelemetryMercuryData, fetcher LatestReportFetcher, linkFeedID, nativeFeedID types.FeedID) *datasource {
+	return &datasource{pr, jb, spec, feedID, lggr, rr, fetcher, linkFeedID, nativeFeedID, sync.RWMutex{}, enhancedTelemChan}
+}
+
+func (ds *datasource) Observe(ctx context.Context, repts ocrtypes.ReportTimestamp, fetchMaxFinalizedTimestamp bool) (obs relaymercuryv2.Observation, merr error) {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		run, trrs, err := ds.executeRun(ctx)
+		if err != nil {
+			merr = fmt.Errorf("Observe failed while executing run: %w", err)
+			obs.BenchmarkPrice.Err = err
+			return
+		}
+		select {
+		case ds.runResults <- run:
+		default:
+			ds.lggr.Warnf("unable to enqueue run save for job ID %d, buffer full", ds.spec.JobID)
+		}
+
+		var parsed parseOutput
+		parsed, err = ds.parse(trrs)
+		if err != nil {
+			// This is not expected under normal circumstances
+			ds.lggr.Errorw("Observe failed while parsing run results", "err", err)
+			merr = fmt.Errorf("Observe failed while parsing run results: %w", err)
+			return
+		}
+		obs.BenchmarkPrice = parsed.benchmarkPrice
+	}()
+
+	// wait for runner to finish
+	wg.Wait()
```

Comment:
Why spawn & wait for a single goroutine instead of executing on this one?
```suggestion
	run, trrs, err := ds.executeRun(ctx)
	if err != nil {
		merr = fmt.Errorf("Observe failed while executing run: %w", err)
		obs.BenchmarkPrice.Err = err
	}  else {
		select {
		case ds.runResults <- run:
		default:
			ds.lggr.Warnf("unable to enqueue run save for job ID %d, buffer full", ds.spec.JobID)
		}
	
		var parsed parseOutput
		parsed, err = ds.parse(trrs)
		if err != nil {
			// This is not expected under normal circumstances
			ds.lggr.Errorw("Observe failed while parsing run results", "err", err)
			merr = fmt.Errorf("Observe failed while parsing run results: %w", err)
		} else {
			obs.BenchmarkPrice = parsed.benchmarkPrice
		}
	}

	var wg sync.WaitGroup
```